dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/DefaultTrackerValidationService.java

```diff
@@ -28,53 +28,114 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-import java.util.ArrayList;
-import java.util.List;
-
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.commons.timer.Timer;
 import org.hisp.dhis.tracker.ValidationMode;
 import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.TrackerValidationHookTimerReport;
 import org.hisp.dhis.tracker.report.TrackerValidationReport;
+import org.hisp.dhis.user.User;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
-import lombok.extern.slf4j.Slf4j;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.hisp.dhis.tracker.validation.TrackerImportValidationConfig.VALIDATION_ORDER;
+import static org.hisp.dhis.tracker.validation.TrackerImportValidationConfig.VALIDATION_ORDER_MAP;
 
 /**
  * @author Morten Olav Hansen <mortenoh@gmail.com>
  */
 @Slf4j
 @Service
-public class DefaultTrackerValidationService implements TrackerValidationService
+public class DefaultTrackerValidationService
+    implements TrackerValidationService
 {
     private List<TrackerValidationHook> validationHooks = new ArrayList<>();
 
     @Autowired( required = false )
     public void setValidationHooks( List<TrackerValidationHook> validationHooks )
     {
         this.validationHooks = validationHooks;
+
+        if ( !VALIDATION_ORDER
```

Comment:
Maybe extracting this logic to a private method with a meaningful name will help understanding what is happening here.

VERY IMPORTANT: We must have a test for this that will fail if such an exception is thrown.
This validation is showing an error in the code, so such codebase should never be deployed

Reply:
Agree this should be in a test!